Kotlin/Wasm: add debugging tutorial (#3964)
JetBrains/kotlin-web-site: docs/topics/wasm/wasm-debugging.md

[//]: # (title: Debug Kotlin/Wasm code)

> Kotlin/Wasm is in [Alpha](components-stability.md). It may be changed at any time.
>
{type="note"}

This tutorial demonstrates how to debug a web application built with Kotlin/Wasm, based on a 
[Compose Multiplatform](https://www.jetbrains.com/lp/compose-multiplatform/) example from our 
`kotlin-wasm-examples` [repository](https://github.com/Kotlin/kotlin-wasm-examples/tree/main).

## Before you start

1. Download and install the latest version of [IntelliJ IDEA](https://www.jetbrains.com/idea/).

> 2. Clone the [Kotlin/Wasm examples](https://github.com/Kotlin/kotlin-wasm-examples/tree/main) repository

Reviewer: ```suggestion
2. Clone the [Kotlin/Wasm examples](https://github.com/Kotlin/kotlin-wasm-examples/) repository
```

   You can also clone it from the command line:

   ```bash
   git clone git@github.com:Kotlin/kotlin-wasm-examples.git
   ```

> Alternatively, you can use our experimental [web wizard](https://kmp.jetbrains.com/) to download a sample project.
>
{type="note"}

## Run the application

1. In IntelliJ IDEA, a **Gradle build scripts found** notification appears. Click **Load**. 

2. Open the **Gradle** tool window by selecting **View** | **Tool Windows** | **Gradle**.

   > You need at least Java 11 as your Gradle JVM for the examples to load successfully.
   >
   {type="note"}

3. In **compose-example** | **Tasks** | **kotlin browser**, select and run the **wasmJsBrowserRun** task.

   ![Run the Gradle task](wasm-gradle-task-window.png){width=650}

    You can also run the following command in the terminal from the `compose-example` directory:

   ```bash
   ./gradlew wasmJsBrowserRun
   ```
   >Alternatively, you can load **compose-example** by opening it directly from the file explorer in **File | Open**.
   >
   {type="note"}

4. Once the application starts, open the following URL in your browser:

   ```bash
   http://localhost:8080/
   ```
   

>    > The port number can vary. If the 8080 port is unavailable, you can find the corresponding port number printed in the console

Reviewer: The actual port number could be different because 8080 is unavailable. You can find the actual port number printed in a Gradle build console.

build file. For more information, see [Configuration for debugging](#configuration-for-debugging).

For this tutorial, we are using Chrome. If you use a different browser, the experience should be similar and the following 
steps should also work. 

To debug a Kotlin/Wasm application:

1. In the browser window of the application, access developer tools by right-clicking and selecting the **Inspect** action.
Alternatively, you can use either the **Fn + F12** or just the  **F12** shortcut.

2. Switch to the **Sources** tab and find the Kotlin file to debug.

3. Set breakpoints in the Kotlin file by clicking on the numbers of the code lines you want to inspect. Only the code lines 
with darker numbers accept breakpoints.

![Set breakpoints](wasm-breakpoints.png){width=700}

4. Interact with the application by clicking on the **Hello World!** button. This action triggers the execution of the 
code, and the debugger pauses at each breakpoint.

5. Inspect variables and code execution at the breakpoints by using the debugging control buttons (step over, step into, step out, and more).

![Debug controls](wasm-debug-controls.png){width=700}

> 6. Check the **Call stack** and **Scope** tool windows to trace the sequence of function calls and pinpoint the location of any errors.

Reviewer: We can do it later

![Check call stack](wasm-debug-scope.png){width=700}

7. Make the required changes to the code and [run the application](#run-the-application) again.

### Configuration for debugging

> In projects requiring additional debugging configuration, you need to allow Webpack servers to access your sources. 

Reviewer: By default, browsers don't have access to your project sources required for debugging. You can configure Webpack DevServer to serve the sources by adding the following code snippet to the Gradle build file:

Add the following code snippet to the Gradle build file:

```kotlin
// Add this line in the dependencies block
import org.jetbrains.kotlin.gradle.targets.js.webpack.KotlinWebpackConfig

// Add these lines inside commonWebpackConfig{}
devServer = (devServer ?: KotlinWebpackConfig.DevServer()).apply {
    static = (static ?: mutableListOf()).apply {
        // Serve sources to debug inside browser
        add(project.projectDir.path)
        add(project.projectDir.path + "/commonMain/")
        add(project.projectDir.path + "/wasmJsMain/")
    }
}
```

> Debugging sources of libraries is currently 
> unavailable. [We are working to address this limitation](https://youtrack.jetbrains.com/issue/KT-64685).
>
{type="note"}

## Leave feedback

* ![Slack](slack.svg){width=25}{type="joined"} Slack: provide your feedback directly to the developers in our [#webassembly](https://kotlinlang.slack.com/archives/CDFP59223) channel. [Get a Slack invite](https://surveys.jetbrains.com/s3/kotlin-slack-sign-up).

> * Report any issues in [YouTrack](https://youtrack.jetbrains.com/issue/KT-56492).

Reviewer: It sounds like call to file an issue, then the link should be different

[//]: # (title: Debug Kotlin/Wasm code)

> Kotlin/Wasm is in [Alpha](components-stability.md). It may be changed at any time.
>
{type="note"}

This tutorial demonstrates how to debug in the browser a [Compose Multiplatform](https://www.jetbrains.com/lp/compose-multiplatform/)
application built with Kotlin/Wasm.

## Before you start

Create a project using the Kotlin Multiplatform wizard:

1. Open the [Kotlin Multiplatform wizard](https://kmp.jetbrains.com/#newProject).

> 2. On the **New Project** tab, you can change the project name to "WasmDemo" and the project ID to "wasm.project.demo".

Author: This is actually optional. They can leave it as it is or add the name as their preference. This change is due to some feedback from the SME saying that it needed to be clarified that this is optional. I modified the step and added the clarification in a separate note. What do you think?

## Open the project in IntelliJ IDEA

1. Download and install the latest version of [IntelliJ IDEA](https://www.jetbrains.com/idea/).
2. On the Welcome screen of IntelliJ IDEA, click **Open** or select **File | Open** in the menu bar.
3. Navigate to the unpacked "WasmDemo" folder and click **Open**.

## Run the application

1. In IntelliJ IDEA, open the **Gradle** tool window by selecting **View** | **Tool Windows** | **Gradle**.

   > You need at least Java 11 as your Gradle JVM for the tasks to load successfully.
   >
   {type="note"}

2. In **composeApp** | **Tasks** | **kotlin browser**, select and run the **wasmJsBrowserRun** task.

   ![Run the Gradle task](wasm-gradle-task-window.png){width=600}

   Alternatively, you can run the following command in the terminal from the `WasmDemo` root directory:

   ```bash
   ./gradlew wasmJsBrowserRun
   ```

3. Once the application starts, open the following URL in your browser:

   ```bash
   http://localhost:8080/
   ```

   >The port number can vary because the 8080 port may be unavailable. You can find the actual port number printed
   > in the Gradle build console.
   >
   {type="tip"}

>    You can see a "Click me!" button. Click it:

Author: Some weeks ago I received feedback from our colleague Aleksey to change this sentence so that it gives more transition to the expected result. I got his point and it made sense to me to change it. His suggestion was also "should". How does it sound to you with "should"?

>    You can see a "Click me!" button. Click it:

Reviewer 2: Interesting! For me, the additional "can" is unnecessary and as a native speaker it somehow infers that it was obscured before in some way. I like to follow [Google's guide](https://developers.google.com/style/word-list#can) in this instance. "Should" is conditional so I would avoid that, for sure! Otherwise it makes it sound like we aren't 100% sure what will happen. I leave the final decision up to you about keeping or removing the "can". 😄 I don't have strong feelings about it.

   ![Click me](wasm-composeapp-browser-clickme.png){width=650}

   Now you can see the Compose Multiplatform logo:

   ![Compose app in browser](wasm-composeapp-browser.png){width=650}

## Debug in the browser

> Currently, debugging is only available in the browser. Debugging in 
> [IntelliJ IDEA](https://youtrack.jetbrains.com/issue/KT-64683/Kotlin-Wasm-debugging-in-IntelliJ-IDEA) and 
> [Fleet](https://youtrack.jetbrains.com/issue/KT-64684) is in the plan. 
>
{type="note"}

You can debug this Compose Multiplatform application
in your browser out of the box without additional configurations. 

However, for other projects, you may need to set the required configuration in the Gradle 

> build file. For more information, see [Configuration for debugging](#configuration-for-debugging).

Reviewer 2: What do you think about moving the "Configuration for debugging" section here but putting the information in a collapsible block? Then the reader doesn't have to jump down and back up the page if they do need to add this additional config 😄

> For this tutorial, we are using Chrome. If you use a different browser, the experience should be similar and the following 
> steps should also work. 
> 
{type="tip"}

To debug a Kotlin/Wasm application:

> 1. In the browser window of the application, access developer tools by right-clicking and selecting the **Inspect** action.

Reviewer 2: Can you change the construction of your instructions for this section so instead of: \<purpose of action\> **by** \<instruction\>
So that you have the instruction first?
For example:
> In the application's browser window, right-click and select 'Inspect' to access the developer tools.

This style is more direct and active.

2. Switch to the **Sources** tab and select the Kotlin file to debug. In this tutorial, we'll work with the `Greeting.kt` file.

3. Set breakpoints in the Kotlin file by clicking on the numbers of the code lines you want to inspect. Only the code lines 
with darker numbers accept breakpoints.

![Set breakpoints](wasm-breakpoints.png){width=700}

4. Interact with the application by clicking on the **Click me!** button. This action triggers the execution of the 
code, and the debugger pauses when the execution reaches a breakpoint.

5. Inspect variables and code execution at the breakpoints by using debugging control buttons such as: 
   * Step into to investigate a function deeper
   * Step over to execute the current line and pause on the next line
   * Step out to execute the code until it exits the current function

![Debug controls](wasm-debug-controls.png){width=700}

6. Check the **Call stack** and **Scope** tool windows to trace the sequence of function calls and pinpoint the location of any errors.

![Check call stack](wasm-debug-scope.png){width=700}

7. Make the required changes to the code and [run the application](#run-the-application) again.

> 

Author: Added. How does it look?

> 

Reviewer 2: Looks great!

> ### Configuration for debugging

Reviewer 2: ```suggestion
### Configure your browser for debugging
```

By default, browsers lack access to the project sources necessary for debugging. To gain access, you can configure the Webpack DevServer 

> to serve these sources. Add the following code snippet to the Gradle build file inside the `ComposeApp` directory:

Reviewer 2: ```suggestion
to serve these sources. In the `ComposeApp` directory, add the following code snippet to your `settings.gradle(.kts)` file:
```

> ```kotlin

Reviewer 2: Can you separate this into two code samples? Then the user can easily copy each snippet into the right place in the file.

// Add this line in the dependencies block
import org.jetbrains.kotlin.gradle.targets.js.webpack.KotlinWebpackConfig

> // Add these lines inside commonWebpackConfig{}

Author: I tried, but then it duplicates some code lines and it's kind of hard to distinguish between the code that already comes in the file and the lines to add; with an extra effort to just copy and paste what is required, avoiding duplication.  I'm thinking of leaving it as it is. Or do you have any suggestions?

> // Add these lines inside commonWebpackConfig{}

Reviewer 2: I think it helps user understand where `commonWebpackConfig{}` is. So for example, something like this:

```
kotlin {
    @OptIn(ExperimentalWasmDsl::class)
    wasmJs {
        moduleName = "composeApp"
        browser {
            commonWebpackConfig {
                outputFileName = "composeApp.js"
                devServer = (devServer ?: KotlinWebpackConfig.DevServer()).apply {
                    static = (static ?: mutableListOf()).apply {
                        // Serve sources to debug inside browser
                        add(project.projectDir.path)
                        add(project.projectDir.path + "/commonMain/")
                        add(project.projectDir.path + "/wasmJsMain/")
                    }
                }
            }
        }
    }
```
Though you'd need to cross-check it with a developer 😄 This is what I commonly do when demonstrating new options in Gradle.

> // Add these lines inside commonWebpackConfig{}

Author: After discussing with the SME, we came up with a solution to clearly show the single `commonWebpackConfig{}` code block they need to add, as well as how it looks inside the whole kotlin{} block.

   ![Click me](wasm-composeapp-browser-clickme.png){width=650}

   Now you see the Compose Multiplatform logo:

   ![Compose app in browser](wasm-composeapp-browser.png){width=650}

## Debug in your browser

> Currently, debugging is only available in your browser. In the future, you will be able to debug your code in 
> [IntelliJ IDEA](https://youtrack.jetbrains.com/issue/KT-64683/Kotlin-Wasm-debugging-in-IntelliJ-IDEA) and 
> [Fleet](https://youtrack.jetbrains.com/issue/KT-64684). 
>
{type="note"}

You can debug this Compose Multiplatform application
in your browser out of the box, without additional configurations. 

However, for other projects, you may need to configure additional settings in your Gradle 
build file. For more information about how to configure your browser for debugging, expand the next section.

### Configure your browser for debugging {initial-collapse-state="collapsed"}

By default, browsers can't access some of the project's sources necessary for debugging. To provide access, you can configure the Webpack DevServer
to serve these sources. In the `ComposeApp` directory, add the following code snippets to your `build.gradle.kts` file.

> Add this line in the dependencies block:

Author: Changed it to "Add this import as a top-level declaration:"

```kotlin
import org.jetbrains.kotlin.gradle.targets.js.webpack.KotlinWebpackConfig
```

> Add this code snippet inside `commonWebpackConfig{}`, located in `kotlin{}` / `wasmJs{}` / `browser{}`:

Reviewer 2: ```suggestion
Add this code snippet inside the `commonWebpackConfig{}` block, located in the `wasmJs{}` target DSL and `browser{}` platform DSL:
```

>
{type="note"}

### Debug your Kotlin/Wasm application

> This tutorial uses the Chrome browser, but you should be able to follow these steps with other browsers. For more information,
> see [Browser versions](wasm-troubleshooting.md#browser-versions).
> 
{type="tip"}

1. In the browser window of the application, right-click and select the **Inspect** action to access developer tools.
   Alternatively, you can use the **F12** shortcut or select **View** | **Developer** | **Developer Tools**.

2. Switch to the **Sources** tab and select the Kotlin file to debug. In this tutorial, we'll work with the `Greeting.kt` file.

3. Click on the line numbers to set breakpoints on the code that you want to inspect. Only the lines
   with darker numbers can have breakpoints.

![Set breakpoints](wasm-breakpoints.png){width=700}

4. Click on the **Click me!** button to interact with the application. This action triggers the execution of the 
   code, and the debugger pauses when the execution reaches a breakpoint.

5. In the Debugging pane, use the debugging control buttons to inspect variables and code execution at the breakpoints:

>    * ![Step into](wasm-step-into.png){width=30}{type="joined"} Step into to investigate a function deeper
>    * ![Step over](wasm-step-over.png){width=30}{type="joined"} Step over to execute the current line and pause on the next line
>    * ![Step out](wasm-step-out.png){width=30}{type="joined"} Step out to execute the code until it exits the current function

Reviewer 2: ```suggestion
   * ![Step into](wasm-step-into.png){width=30}{type="joined"} Step into, to investigate a function more deeply.
   * ![Step over](wasm-step-over.png){width=30}{type="joined"} Step over, to execute the current line and pause on the next line.
   * ![Step out](wasm-step-out.png){width=30}{type="joined"} Step out, to execute the code until it exits the current function.
```

To debug a Kotlin/Wasm application:

1. In the browser window of the application, access developer tools by right-clicking and selecting the **Inspect** action.

> Alternatively, you can use the **Fn + F12** shortcut.

Reviewer: I believe it's just `F12`. In your case, you probably need to press `Fn` due to default to media buttons.
